Remotion/Data/DomainObjects/Persistence/Rdbms/SqlServer/SqlDialect.cs

```diff
@@ -50,5 +53,61 @@ public virtual string DelimitIdentifier (string identifier)
       return "[" + identifier + "]";
     }
 
+    public IDbDataParameter CreateDataParameter (IDbCommand command, IStorageTypeInformation storageTypeInformation, string parameterName, object? value)
+    {
+      ArgumentUtility.CheckNotNull("command", command);
+      ArgumentUtility.CheckNotNull("storageTypeInformation", storageTypeInformation);
+      ArgumentUtility.CheckNotNullOrEmpty("parameterName", parameterName);
+
+      var convertedValue = storageTypeInformation.ConvertToStorageType(value);
+
+      var parameter = command.CreateParameter();
+      parameter.ParameterName = GetParameterName(parameterName);
+      parameter.Value = convertedValue;
+      parameter.DbType = storageTypeInformation.StorageDbType;
+      if (storageTypeInformation.StorageTypeLength.HasValue)
+      {
+        var parameterSize = storageTypeInformation.StorageTypeLength.Value;
+        if (parameterSize == SqlStorageTypeInformationProvider.StorageTypeLengthRepresentingMax)
+        {
+          var isAnsiString = storageTypeInformation.StorageDbType == DbType.AnsiString; //AnsiStringFixedLength represents char(...), which does not support char(max)
+          var isUnicodeString = storageTypeInformation.StorageDbType == DbType.String; //StringFixedLength represents char(...), which does not support char(max)
+
+          int? fulltextCompatibleMaxLength = null;
+          if (isAnsiString)
+            fulltextCompatibleMaxLength = 8000;
+
+          if (isUnicodeString)
+            fulltextCompatibleMaxLength = 4000;
+
+          if (fulltextCompatibleMaxLength.HasValue)
+          {
+            var isNullValue = parameter.Value == null || parameter.Value == DBNull.Value;
+            var isStringAndValueDoesNotExceedCompatibleMaxLength =
+                parameter.Value is string stringValue && stringValue.Length <= fulltextCompatibleMaxLength;
+            var isCharArrayAndValueDoesNotExceedCompatibleMaxLength =
+                parameter.Value is char[] charsValue && charsValue.Length <= fulltextCompatibleMaxLength;
+
+            if (isNullValue || isStringAndValueDoesNotExceedCompatibleMaxLength || isCharArrayAndValueDoesNotExceedCompatibleMaxLength)
+              parameter.Size = fulltextCompatibleMaxLength.Value;
```

Comment:
We could simplify to the following.
```suggestion
            var isStringAndValueDoesNotExceedCompatibleMaxLength =
            bool isTextAndValueDoesNotExceedCompatibleMaxLength = dbDataParameter.Value switch
            {
                string stringValue => stringValue.Length <= _fulltextCompatibleMaxLength,
                char[] charsValue => charsValue.Length <= _fulltextCompatibleMaxLength,
                _ => false
            };
            if (isNullValue || isTextAndValueDoesNotExceedCompatibleMaxLength)
              parameter.Size = fulltextCompatibleMaxLength.Value;
```

I also considered using this just to get "Length or null" from the DataParameter.Value, but it felt easier to read this way. the boolean test in the IF is clearer that way / naming the variables would be awkward. I do accept good version that do not have the comparison twice, though.

Actually, we can integrate the null test into the switch-pattern too and make it one happy boolean value which would look like this:
```
      bool hasFulltextCompatibleMaxLength = dbDataParameter.Value switch
      {
          null => true,
          DBNull => true,
          string stringValue => stringValue.Length <= _fulltextCompatibleMaxLength,
          char[] charsValue => charsValue.Length <= _fulltextCompatibleMaxLength,
          _ => false
      };
      if (hasFulltextCompatibleMaxLength)
        parameter.Size = fulltextCompatibleMaxLength.Value;
```
Which one is best really depends on whether you wanto be explicit about the null-behavior in the if condition.